src/docker.rs

```diff
@@ -107,29 +109,78 @@ fn get_docker_for_tcp(tcp_host_str: String) -> Docker {
     }
 }
 
-// https://docs.docker.com/reference/api/docker_remote_api_v1.17/
+#[cfg(feature = "tls")]
+fn get_docker_for_tcp_tls(
+    host: String,
+    cert_path: &Path,
+    verify: bool,
+) -> Result<Docker> {
+    let http = get_http_connector();
+    let mut connector = SslConnector::builder(SslMethod::tls())?;
+    connector.set_cipher_list("DEFAULT")?;
+    let cert = cert_path.join("cert.pem");
+    let key = cert_path.join("key.pem");
+    connector.set_certificate_file(cert.as_path(), SslFiletype::PEM)?;
+    connector.set_private_key_file(key.as_path(), SslFiletype::PEM)?;
+    if verify {
+        let ca = cert_path.join("ca.pem");
+        connector.set_ca_file(ca.as_path())?;
+    }
+
+    Ok(Docker {
+        transport: Transport::EncryptedTcp {
+            client: Client::builder().build(HttpsConnector::with_connector(http, connector)?),
+            host: format!("https://{}", host),
+        },
+    })
+}
+
 impl Docker {
-    /// constructs a new Docker instance for a docker host listening at a url specified by an env var `DOCKER_HOST`,
-    /// falling back on unix:///var/run/docker.sock
-    pub fn new() -> Docker {
-        match env::var("DOCKER_HOST").ok() {
-            Some(host) => {
-                #[cfg(feature = "unix-socket")]
-                if let Some(path) = host.strip_prefix("unix://") {
-                    return Docker::unix(path);
+    /// Creates a new Docker instance by automatically choosing appropriate connection type based
+    /// on provided `uri`.
+    ///
+    /// Supported schemes are:
+    ///  - `unix://` with feature `unix-socket` enabled, otherwise returns an Error
+    ///  - `tcp://`
+    ///  - `http://`
+    ///
+    ///  To create a Docker instance utilizing TLS use explicit [Docker::tls](Docker::tls)
+    ///  constructor.
+    pub fn new<S>(uri: S) -> Result<Docker>
```

Comment:
I am not sure whether I like the automatic finding of the scheme here, because a user must rely on the correct implementation of the parsing.
Maybe we can have something like this:

```
Docker::new(DockerConnectionScheme::from_str("unix://...")?)
```

So that we keep this general idea, but also give the user the option to be explicit, be having him construct a `DockerConnectionScheme` themselves:

```
pub enum DockerConnectionScheme {
    UnixSocket(PathBuf),
    Tcp(url::Url), // or similar - strongly typed
    Http(url::Url), //...
}
```

And with an appropriate `std::str::FromStr` implementation that does the parsing. This way a user can chose what they want!

What do you think?

---

(Above code is just a quick mockup of my idea, details may be different)